File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int curr = 0;
        int prev = 0;
        int next = 0;
        int sz = nums.size();
        
        for(int i=0; i<nums.size()-1;i++){
            next = std::max(prev + nums[i], curr);
            prev = curr;
            curr = next;
        }
        int rob1 = curr;
        prev = 0;
        curr = 0;
        next = 0;
        for(int i=1; i<nums.size();i++){
            next = std::max(prev + nums[i], curr);
            prev = curr;
            curr = next;
        }
        int rob2 = curr;
        
        return std::max(nums[0],std::max(rob1,rob2));
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]


        def simple_rob(nums):
            n = len(nums)
            print(nums)
            if len(nums) == 1:
                return nums[0]
            prev, curr = nums[0], max(nums[0], nums[1])

            for i in range(2,n):

                prev, curr = curr, max(curr, prev + nums[i])

            return curr

        return max(simple_rob(nums[:-1]), simple_rob(nums[1:]))